// jsxapi/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.connect = exports.connectGen = exports.XAPI = void 0;
var ws_1 = __importDefault(require("ws"));
var tsh_1 = __importDefault(require("./backend/tsh"));
var ws_2 = __importDefault(require("./backend/ws"));
var connect_1 = __importDefault(require("./connect"));
var ssh_1 = __importDefault(require("./transport/ssh"));
var tsh_2 = __importDefault(require("./transport/tsh"));
var ws_3 = __importDefault(require("./transport/ws"));
var xapi_1 = __importDefault(require("./xapi"));
var xapi_2 = require("./xapi");
Object.defineProperty(exports, "XAPI", { enumerable: true, get: function () { return __importDefault(xapi_2).default; } });
function initBackend(opts) {
    var host = opts.host, port = opts.port, protocol = opts.protocol;
    switch (protocol) {
        case '':
        case 'ssh:': {
            var transport = (0, ssh_1.default)(opts);
            return new tsh_1.default(transport);
        }
        case 'tsh:': {
            var transport = (0, tsh_2.default)(host, port);
            return new tsh_1.default(transport);
        }
        case 'ws:':
        case 'wss:': {
            var createWebSocket = function (url, auth) {
                var ws = new ws_1.default(url, auth, {
                    followRedirects: true,
                    rejectUnauthorized: false,
                });
                return ws;
            };
            var transport = (0, ws_3.default)(createWebSocket, opts);
            return new ws_2.default(transport);
        }
        default:
            throw new Error("Invalid protocol: " + protocol);
    }
}
function connectGen(xapi) {
    return (0, connect_1.default)(initBackend, { protocol: 'wss:' })(xapi);
}
exports.connectGen = connectGen;
exports.connect = connectGen(xapi_1.default);
