// jsxapi/lib/xapi/exc.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterError = exports.InvalidPathError = exports.IllegalValueError = exports.XAPIError = exports.METHOD_NOT_FOUND = exports.INVALID_STATUS = exports.INVALID_RESPONSE = exports.PARAMETER_ERROR = exports.INVALID_PATH = exports.ILLEGAL_VALUE = exports.COMMAND_ERROR = exports.UNKNOWN_ERROR = void 0;
exports.UNKNOWN_ERROR = 0;
exports.COMMAND_ERROR = 1;
exports.ILLEGAL_VALUE = 2;
exports.INVALID_PATH = 3;
exports.PARAMETER_ERROR = 4;
exports.INVALID_RESPONSE = 5;
exports.INVALID_STATUS = 6;
exports.METHOD_NOT_FOUND = -32601;
var XAPIError = (function (_super) {
    __extends(XAPIError, _super);
    function XAPIError(code, reason, data) {
        var _this = _super.call(this, reason) || this;
        _this.code = code;
        Object.setPrototypeOf(_this, XAPIError.prototype);
        if (data !== undefined) {
            _this.data = data;
        }
        if (typeof reason !== 'string') {
            throw new Error('Reason for XAPIError must be a string');
        }
        if (typeof code !== 'number') {
            throw new Error('Error code for XAPIError must be a number');
        }
        return _this;
    }
    return XAPIError;
}(Error));
exports.XAPIError = XAPIError;
var IllegalValueError = (function (_super) {
    __extends(IllegalValueError, _super);
    function IllegalValueError(reason) {
        var _this = _super.call(this, exports.ILLEGAL_VALUE, reason) || this;
        Object.setPrototypeOf(_this, IllegalValueError.prototype);
        return _this;
    }
    return IllegalValueError;
}(XAPIError));
exports.IllegalValueError = IllegalValueError;
var InvalidPathError = (function (_super) {
    __extends(InvalidPathError, _super);
    function InvalidPathError(reason, xpath) {
        var _this = _super.call(this, exports.INVALID_PATH, reason, { xpath: xpath }) || this;
        Object.setPrototypeOf(_this, InvalidPathError.prototype);
        return _this;
    }
    return InvalidPathError;
}(XAPIError));
exports.InvalidPathError = InvalidPathError;
var ParameterError = (function (_super) {
    __extends(ParameterError, _super);
    function ParameterError() {
        var _this = _super.call(this, exports.PARAMETER_ERROR, 'Invalid or missing parameters') || this;
        Object.setPrototypeOf(_this, ParameterError.prototype);
        return _this;
    }
    return ParameterError;
}(XAPIError));
exports.ParameterError = ParameterError;
