// jsxapi/lib/xapi/mixins.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Settable = exports.Gettable = exports.Listenable = exports.Component = void 0;
var normalizePath_1 = __importDefault(require("./normalizePath"));
var Component = (function () {
    function Component(xapi) {
        this.xapi = xapi;
        this.prefix = '';
    }
    Component.prototype.normalizePath = function (path) {
        var normalized = (0, normalizePath_1.default)(path);
        var prefix = this.prefix;
        return !prefix
            ? normalized
            : [prefix].concat(normalized);
    };
    return Component;
}());
exports.Component = Component;
function Listenable(Base) {
    return (function (_super) {
        __extends(Child, _super);
        function Child() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Child.prototype.on = function (path, listener) {
            return this.xapi.feedback.on(this.normalizePath(path), listener);
        };
        Child.prototype.once = function (path, listener) {
            return this.xapi.feedback.once(this.normalizePath(path), listener);
        };
        Child.prototype.off = function () {
            this.xapi.feedback.off();
        };
        return Child;
    }(Base));
}
exports.Listenable = Listenable;
function Gettable(Base) {
    return (function (_super) {
        __extends(Child, _super);
        function Child() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Child.prototype.get = function (path) {
            return this.xapi.execute('xGet', {
                Path: this.normalizePath(path),
            });
        };
        return Child;
    }(Base));
}
exports.Gettable = Gettable;
function Settable(Base) {
    return (function (_super) {
        __extends(Child, _super);
        function Child() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Child.prototype.set = function (path, value) {
            return this.xapi.execute('xSet', {
                Path: this.normalizePath(path),
                Value: value,
            });
        };
        return Child;
    }(Base));
}
exports.Settable = Settable;
